pyro/infer/abstract_infer.py

```diff
@@ -2,133 +2,63 @@
-    __call__ = sample
-
-    def log_prob(self, val, *args, **kwargs):
-        d, values = self._dist_and_values(*args, **kwargs)
-        ix = _index(values, val)
-        return d.log_prob(torch.tensor([ix]))
-
-    def enumerate_support(self, *args, **kwargs):
-        d, values = self._dist_and_values(*args, **kwargs)
-        return values[:]
-
-
-class Marginal(Histogram):
-    """
-    :param trace_dist: a TracePosterior instance representing a Monte Carlo posterior
-
-    Marginal histogram distribution.
-    Turns a TracePosterior object into a Distribution
-    over the return values of the TracePosterior's model.
-    """
     def __init__(self, trace_dist, sites=None):
         assert isinstance(trace_dist, TracePosterior), \
             "trace_dist must be trace posterior distribution object"
 
         if sites is None:
-            sites = "_RETURN"
-
-        assert isinstance(sites, (str, list)), \
-            "sites must be either '_RETURN' or list"
-
-        if isinstance(sites, str):
-            assert sites in ("_RETURN",), \
-                "sites string must be '_RETURN'"
+            sites = ["_RETURN"]
 
         self.sites = sites
-        super(Marginal, self).__init__()
         self.trace_dist = trace_dist
+        self.empirical_marginal = None
+        super(Distribution, self).__init__()
+
+    def sample(self, *args, **kwargs):
+        sample_shape = kwargs.pop("sample_shape", torch.Size())
+        return self._get_marginal(*args, **kwargs).sample(sample_shape)
+
+    def log_prob(self, value, *args, **kwargs):
+        return self._get_marginal(*args, **kwargs).log_prob(value)
+
+    def enumerate_support(self, *args, **kwargs):
+        return self._get_marginal(*args, **kwargs).enumerate_support()
 
-    def _gen_weighted_samples(self, *args, **kwargs):
-        for tr, log_w in poutine.block(self.trace_dist._traces)(*args, **kwargs):
-            if self.sites == "_RETURN":
-                val = tr.nodes["_RETURN"]["value"]
-            else:
-                val = {name: tr.nodes[name]["value"]
-                       for name in self.sites}
-            yield (val, log_w)
+    @util.memoize
+    def _get_marginal(self, *args, **kwargs):
+        return self.trace_dist(*args, **kwargs).marginal(self.sites)
 
 
 class TracePosterior(object):
     """
     Abstract TracePosterior object from which posterior inference algorithms inherit.
-    Holds a generator over Traces sampled from the approximate posterior.
-    Not actually a distribution object - no sample or score methods.
+    When called, collects a bag of execution traces from the approximate posterior.
+    Not itself a distribution object, but provides convenience methods that return
+    an empirical distribution.
     """
     def __init__(self):
-        pass
+        self._init()
+
+    def _init(self):
+        self.log_weights = []
+        self.exec_traces = []
+        self._marginals = {}
 
     def _traces(self, *args, **kwargs):
```

Comment:
If this is still getting used externally we should remove `_`

Reply:
The base class raises the exception so that the inheriting class overrides it. This isn't part of the API, and with EmpiricalMarginal, we shouldn't have to ever call `_traces`. If we have any such use cases, we will incorporate them into helper methods/functions.